File: web/encodings.js
const TRUE = parse("λx.λy.x");
const FALSE = parse("λx.λy.y");

const AND = parse("λp.λq.p q p");
const OR = parse("λp.λq.p p q");
const NOT = parse("λp.λa.λb.p b a");
const IF_THEN_ELSE = parse("λp.λa.λb.p a b");

const ZERO = parse("λf.λx.x");
const ONE = parse("λf.λx.f x");
const TWO = parse("λf.λx.f (f x)");
const THREE = parse("λf.λx.f (f (f x))");

const SUCC = parse("λn.λf.λx.f (n f x)");
const PLUS = parse("λm.λn.λf.λx.m f (n f x)");
const MULT = parse("λm.λn.λf.m (n f)");
const POW = parse("λm.λn.n m");

const PRED = parse("λn.λf.λx.n (λg.λh.h (g f)) (λu.x) (λu.u)");
const SUB = parse("λm.λn.n PRED m");
const IS_ZERO = parse("λn.n (λx.FALSE) TRUE");
const LEQ = parse("λm.λn.IS_ZERO (SUB m n)")
const EQ = parse("λm.λn.AND (LEQ m n) (LEQ n m)");

const PAIR = parse("λx.λy.λf.f x y");
const FIRST = parse("λp.p (λx.λy.x)");
const SECOND = parse("λp.p (λx.λy.y)");

const NIL = parse("λx.TRUE");
const IS_NIL = parse("λl.l (λh.λt.λd.FALSE)");
const CONS = parse("λh.λt.λc.c h t");
const HEAD = parse("λl.l (λh.λt.h)");
const TAIL = parse("λl.l (λh.λt.t)");

function churchNumeral(n) {
    if (n < 0) {
        throw new Error("Church numerals only represent natural numbers");
    }
    
    if (n === 0) {
        return ZERO;
    }
    
    let result = ZERO;
    for (let i = 0; i < n; i++) {
        result = new Application(SUCC, result);
    }
    
    return result;
}

function extractChurchNumeral(expr) {
    if (!(expr instanceof Abstraction) || !(expr.body instanceof Abstraction)) {
        return null;
    }
    
    const fParam = expr.param;
    const xParam = expr.body.param;
    let body = expr.body.body;
    
    let count = 0;
    let current = body;
    
    while (current instanceof Application) {
        if (!(current.left instanceof Variable) || current.left.name !== fParam) {
            return null;
        }
        
        current = current.right;
        count++;
    }
    
    if (!(current instanceof Variable) || current.name !== xParam) {
        return null;
    }
    
    return count;
}

function toBoolean(expr) {
    const [reduced, _, __] = reduceToNormalForm(expr);
    
    if (reduced.isAlphaEquivalent(TRUE)) {
        return true;
    } else if (reduced.isAlphaEquivalent(FALSE)) {
        return false;
    } else {
        return null;
    }
}

function createChurchList(items, encoderFn = null) {
    let result = NIL;
    
    for (let i = items.length - 1; i >= 0; i--) {
        const item = items[i];
        const encodedItem = encoderFn ? encoderFn(item) : item;
        
        result = new Application(
            new Application(CONS, encodedItem),
            result
        );
    }
    
    return result;
}

File: lambda_calculus/encodings.py
from lambda_calculus.syntax import Variable, Abstraction, Application
from lambda_calculus.parser import parse

TRUE = parse("λx.λy.x")          
FALSE = parse("λx.λy.y")         

AND = parse("λp.λq.p q p")       
OR = parse("λp.λq.p p q")        
NOT = parse("λp.λa.λb.p b a")    
IF_THEN_ELSE = parse("λp.λa.λb.p a b")  

ZERO = parse("λf.λx.x")            
ONE = parse("λf.λx.f x")           
TWO = parse("λf.λx.f (f x)")       
THREE = parse("λf.λx.f (f (f x))") 

SUCC = parse("λn.λf.λx.f (n f x)")     
PLUS = parse("λm.λn.λf.λx.m f (n f x)") 
MULT = parse("λm.λn.λf.m (n f)")        
POW = parse("λm.λn.n m")                
PRED = parse("λn.λf.λx.n (λg.λh.h (g f)) (λu.x) (λu.u)")
SUB = parse("λm.λn.n PRED m")

IS_ZERO = parse("λn.n (λx.FALSE) TRUE")  
LEQ = parse("λm.λn.IS_ZERO (SUB m n)")   
EQ = parse("λm.λn.AND (LEQ m n) (LEQ n m)")


PAIR = parse("λx.λy.λf.f x y")     
FIRST = parse("λp.p (λx.λy.x)")    
SECOND = parse("λp.p (λx.λy.y)")   


NIL = parse("λx.TRUE")               
IS_NIL = parse("λl.l (λh.λt.λd.FALSE)") 
CONS = parse("λh.λt.λc.c h t")       
HEAD = parse("λl.l (λh.λt.h)")       
TAIL = parse("λl.l (λh.λt.t)")       


def church_numeral(n):
    if n < 0:
        raise ValueError("Church numerals only represent natural numbers")
    
    if n == 0:
        return ZERO
    
    result = ZERO
    for _ in range(n):
        result = Application(SUCC, result)
    
    return result


def extract_church_numeral(expr):
    if not isinstance(expr, Abstraction) or not isinstance(expr.body, Abstraction):
        return None
    
    f_param = expr.param
    x_param = expr.body.param
    body = expr.body.body
    
    count = 0
    current = body
    
    while isinstance(current, Application):
        if not isinstance(current.left, Variable) or current.left.name != f_param:
            return None
        
        current = current.right
        count += 1
    
    if not isinstance(current, Variable) or current.name != x_param:
        return None
    
    return count


def to_boolean(expr):
    from lambda_calculus.semantics import reduce_to_normal_form
    
    reduced, _, _ = reduce_to_normal_form(expr)
    if reduced.is_alpha_equivalent(TRUE):
        return True
    elif reduced.is_alpha_equivalent(FALSE):
        return False
    else:
        return None


def create_church_list(items, encoder_fn=None):
    result = NIL
    
    for item in reversed(items):
        if encoder_fn:
            encoded_item = encoder_fn(item)
        else:
            encoded_item = item
        
        result = Application(
            Application(CONS, encoded_item),
            result
        )
    
    return result
